pgx/transaction.go

```diff
@@ -0,0 +1,139 @@
+//go:build go1.16
+// +build go1.16
+
+package pgx
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/jackc/pgx/v4"
+	"go.uber.org/multierr"
+
+	trmsql "github.com/avito-tech/go-transaction-manager/sql"
+	"github.com/avito-tech/go-transaction-manager/trm"
+)
+
+// Transaction is trm.Transaction for pgx.Tx.
+type Transaction struct {
+	tx        pgx.Tx
+	savePoint trmsql.SavePoint
+	saves     int64
+	isActive  int64
+}
+
+// NewTransaction creates trm.Transaction for pgx.Tx.
+func NewTransaction(
+	ctx context.Context,
+	sp trmsql.SavePoint,
+	txOptions *pgx.TxOptions,
+	db Transactional,
+) (context.Context, *Transaction, error) {
+	var opts pgx.TxOptions
+	if txOptions != nil {
+		opts = *txOptions
+	}
+
+	tx, err := db.BeginTx(ctx, opts)
+	if err != nil {
+		return ctx, nil, err
+	}
+
+	tr := &Transaction{
+		tx:        tx,
+		savePoint: sp,
+		isActive:  1,
+		saves:     0,
+	}
+
+	go tr.awaitDone(ctx)
+
+	return ctx, tr, nil
+}
+
+func (t *Transaction) awaitDone(ctx context.Context) {
+	if ctx.Done() == nil {
+		return
+	}
+
+	<-ctx.Done()
+
+	t.deactivate()
+}
+
+// Transaction returns the real transaction pgx.Tx.
+func (t *Transaction) Transaction() interface{} {
+	return t.tx
+}
+
+// Begin nested transaction by save point.
+func (t *Transaction) Begin(ctx context.Context, _ trm.Settings) (context.Context, trm.Transaction, error) { //nolint:ireturn,nolintlint
+	_, err := t.tx.Exec(ctx, t.savePoint.Create(t.incrementID()))
```

Comment:
pgx has own implementation of savePoint, I think we can rely them.

`t.tx.Begin(ctx)`

Reply:
Done.
Also modify unittests due to these changes